### ISSUE TITLE
[feature request] Combining with flash attention?

### ISSUE THREAD

**Issue by Vbansal21** [lucidrains/memory-efficient-attention-pytorch#2]
There is a new algorithm to optimize the qkv attention, 
https://github.com/HazyResearch/flash-attention 
https://arxiv.org/abs/2205.14135
It optimises the qkv attention part.
Maybe you can look into integrating it with this.

**Comment by lucidrains** [comment]
@Vbansal21 yup, will do! just wrapping up some denoising diffusion code

**Comment by Vbansal21** [comment]
Thanks 🥳🥳🥳.
I wonder, how do you manage to do all that‽‽🤯🤯🤯

**Comment by lucidrains** [comment]
coffee

**Comment by lucidrains** [comment]
probably won't build it until i get a response from the authors https://github.com/HazyResearch/flash-attention/issues/10

**Comment by Vbansal21** [comment]
Sure, take your time.😀👍

**Comment by KhrulkovV** [comment]
Hey!
Tried FlashAttention in my DDPM training code (CIFAR-10), the loss is behaving a bit unstable indeed
<img width="287" alt="Screen Shot 2022-06-16 at 14 03 13" src="https://user-images.githubusercontent.com/7949649/174056694-8cf37161-62d0-48c6-b6cf-7dc28a9b9839.png">
So something suspicious is probably going on

**Comment by lucidrains** [comment]
@KhrulkovV Hi Valentin, thanks for sharing that

what sequence length and column block size are you using? and do you have a non-flash attention run for comparison?

**Comment by lucidrains** [comment]
yea, they haven't responded, it is not a great sign

i'm just going to assume it doesn't work

**Comment by Vbansal21** [comment]
> yea, they haven't responded, it is not a great sign
> 
> i'm just going to assume it doesn't work

Sure thing.👍
Thanks for considering.😁

**Comment by lucidrains** [comment]
@KhrulkovV did you train using dropout or no?

**Comment by KhrulkovV** [comment]
> @KhrulkovV did you train using dropout or no?

I don’t remember tbh, now I am training with dropout 0.15-0.3 and got fid 2.7

**Comment by lucidrains** [comment]
@KhrulkovV oh wow, is this with the most recent changes to the repository?

**Comment by lucidrains** [comment]
@KhrulkovV Tri added bf16 support to flash attention just yesterday, and that should further help with numerical stability

**Comment by lucidrains** [comment]
i was wrong. flash attention may be the real deal! i'll be playing around with it soon and maybe build out a jax version for practice

**Comment by KhrulkovV** [comment]
> @KhrulkovV oh wow, is this with the most recent changes to the repository?

I actually yeeted unet from guided-diffusion repo for these experiments as I wanted to replicate the paper exactly ;(

**Comment by KhrulkovV** [comment]
> @KhrulkovV Tri added bf16 support to flash attention just yesterday, and that should further help with numerical stability

Training in bf16 for some reason is extremely slow for me but I did not experiment with it extensively. Still fighting the color shift problem
